test/types/array.js

```diff
@@ -578,13 +578,73 @@ describe('array', () => {
 
         it('can fill the default values into the value array', () => {
 
-            const schema = Joi.array().ordered( Joi.string().required(), Joi.number().default(0), Joi.number().default(6)).required();
+            const schema = Joi.array().ordered(Joi.string().required(), Joi.number().default(0), Joi.number().default(6)).required();
 
             Helper.validate(schema, [
                 [['a'], true, ['a', 0, 6]]
             ]);
         });
 
+        it('ignore trailing undefined', () => {
+
+            const schema = Joi.array().ordered(Joi.string().required(), Joi.string(), Joi.number().default(5), Joi.string(), Joi.string());
+
+            Helper.validate(schema, [
+                [['a'], true, ['a', undefined, 5]]
+            ]);
+        });
+
+        it('generate sparse array', () => {
+
+            const schema = Joi.array().ordered(Joi.string(), Joi.number().default(5), Joi.string(), Joi.string().default('f'));
+
+            Helper.validate(schema, [
+                [[], true, [undefined, 5, undefined, 'f']]
+            ]);
+        });
+
+        it('fill the defaults correctly when dynamic schema is used', () => {
```

Comment:
Change the test title to `fills defaults correctly when nested items contain references` to be more meaningful